synapse/lib/view.py

```diff
@@ -138,12 +119,263 @@ async def __anit__(self, core, node):
+    def reqParentQuorum(self):
+
+        if self.parent is None:
+            mesg = f'View ({self.iden}) has no parent.'
+            raise s_exc.SynErr(mesg=mesg)
+
+        quorum = self.parent.info.get('quorum')
+        if quorum is None:
+            mesg = f'Parent view of ({self.iden}) does not require quorum voting.'
+            raise s_exc.SynErr(mesg=mesg)
+
+        return quorum
+
+    def reqNoParentQuorum(self):
+
+        if self.parent is None:
+            return
+
+        quorum = self.parent.info.get('quorum')
+        if quorum is not None:
+            mesg = f'Parent view of ({self.iden}) requires quorum voting.'
+            raise s_exc.SynErr(mesg=mesg)
+
+    async def _wipeViewMeta(self):
+        for lkey in self.core.slab.scanKeysByPref(self.bidn, db='view:meta'):
+            self.core.slab.delete(lkey, db='view:meta')
+            await asyncio.sleep(0)
+
+    def getMergeRequest(self):
+        byts = self.core.slab.get(self.bidn + b'merge:req', db='view:meta')
+        if byts is not None:
+            return s_msgpack.un(byts)
+
+    async def setMergeRequest(self, mergeinfo):
+        self.reqParentQuorum()
+        mergeinfo['iden'] = s_common.guid()
+        mergeinfo['created'] = s_common.now()
+        return await self._push('merge:set', mergeinfo)
+
+    @s_nexus.Pusher.onPush('merge:set')
+    async def _setMergeRequest(self, mergeinfo):
+        self.reqParentQuorum()
+        s_schemas.reqValidMerge(mergeinfo)
+        lkey = self.bidn + b'merge:req'
+        self.core.slab.put(lkey, s_msgpack.en(mergeinfo), db='view:meta')
+        return mergeinfo
+
+    @s_nexus.Pusher.onPushAuto('merge:del')
+    async def delMergeRequest(self):
+        self.reqParentQuorum()
+        byts = self.core.slab.pop(self.bidn + b'merge:req', db='view:meta')
+
+        await self._delMergeMeta()
+
+        if byts is not None:
+            return s_msgpack.un(byts)
+
+    async def _delMergeMeta(self):
+        for lkey in self.core.slab.scanKeysByPref(self.bidn + b'merge:', db='view:meta'):
+            await asyncio.sleep(0)
+            self.core.slab.delete(lkey, db='view:meta')
+
+    async def getMergeVotes(self):
+        for lkey, byts in self.core.slab.scanByPref(self.bidn + b'merge:vote', db='view:meta'):
+            await asyncio.sleep(0)
+            yield s_msgpack.un(byts)
+
+    async def getMerges(self):
+        '''
+        Yield the historical merges into this view.
+        '''
+        for lkey, bidn in self.core.slab.scanByPrefBack(self.bidn + b'hist:merge:time', db='view:meta'):
+            byts = self.core.slab.get(self.bidn + b'hist:merge:iden' + bidn, db='view:meta')
+            if byts is not None:
+                yield s_msgpack.un(byts)
+            await asyncio.sleep(0)
+
+    async def tryToMerge(self, tick):
+        # NOTE: must be called from within a nexus handler!
+
+        if self.merging:
+            return
+
+        if not await self.isMergeReady():
+            return
+
```

Comment:
maybe a call to `mergeAllowed()` here? What if the merge request was opened several days ago and since then someone locked the parent's top layer? That wouldn't be handled gracefully.

Reply:
no. mergeAllowed() checks individual edit permissions vs the users granted permissions and is not applicable to quorum vote based merges

Reply:
( i'll add checking for parent top layer readonly status though )